README.md
# ibm-service-validator 0.5.0

Extends [Schemathesis](https://github.com/kiwicom/schemathesis) to test [IBM API Handbook](https://cloud.ibm.com/docs/api-handbook?topic=api-handbook-intro) compliance and consistency between an API implementation and its OpenAPI definition.

## Overview

This tool takes an OpenAPI definition, a valid API endpoint, and any necessary API credentials and verifies that the API implementation complies with its OpenAPI definition and the [IBM API Handbook](https://cloud.ibm.com/docs/api-handbook?topic=api-handbook-intro).

## Python Version

The minimum python version supported is 3.5.8

## Install

The Service Validator is distributed through PyPI, and you may install using `pip`.

    pip3 install ibm-service-validator

## Use

### Run

The `run` command runs a suite of tests against the service using an API definition.

    ibm-service-validator [env] run <API definition> --base-url <base URL of API> [run options]

#### [env]

The following options may be used to set environment variables used to obtain a bearer token when --with-bearer used:

- --set-api-key: set the IBM_CLOUD_SERVICE_VALIDATOR_API_KEY environment variable to the API key needed to obtain a bearer token.
- --set-iam-endpoint: set the IBM_CLOUD_SERVICE_VALIDATOR_IAM_ENDPOINT to the URL of the API at which the bearer token may be obtained.

Example Usage:

    ibm-service-validator --set-api-key=YOUR_API_KEY --set-iam-endpoint=https://iam.test.cloud.ibm.com/identity/token run path/to/schema -b https://api.com --with-bearer

#### run options

- -a (--auth): provide server username and password in the form `username:password`.
- -A (--auth-type): authentication mechanism. May be "basic" or "digest" (default is "basic").
- -b (--base-url): base url of the service to be tested.
- -c (--checks): comma-separated list of checks to run. Example: `--checks=not_a_server_error,response_schema_conformance`.
  - See [configuration](#configuration) for full list of checks
- -x (--exitfirst): flag to exit and report on the first error or test failure.
- -H (--header): custom header to include in all requests. Example: `-H Authorization:Bearer\ 123`.
- -v (--verbosity): increase the verbosity of the report using the repetition of options. Examples: `-v`, `-vv`, `-vvv` in order of increasing verbosity. We only use one level of verbosity but this is passed to schemathesis which may utilize more levels of verbosity.
- -B (--with-bearer): obtains a bearer token and includes it in tests. Uses [environment variables](#env) to obtain the bearer token.
- --show-errors-tracebacks: flag to show error tracebacks for internal errors.
- --store-request-log: name of yaml file in which to store logs of requests made during testing. Example: `--store-request-log=logs.yaml`.
- --hypothesis-deadline: number of milliseconds allowed for the server to respond (default is 500). Example: `--hypothesis-deadline=300`.
- --hypothesis-phases: determines how test data will be generated. **The default value, `explicit`, indicates test data will only be generated from examples in the OpenAPI definition.** Example: `--hypothesis-phases=explicit,generate` will use explicit OpenAPI examples and generate test data.
  - `explicit`: test data generated from examples. Recommended.
  - `reuse`: reuse old test data.
  - `generate`: test data generated from schema definition.
  - `target`: mutate test data for targeting.
  - `shrink`: shrink test data.
- --hypothesis-verbosity: control how much information is reported. Example: `--hypothesis-verbosity=verbose`
  - `quiet`
  - `normal`
  - `verbose`
  - `debug`

Filter tests by endpoints, methods, and/or tags that match a given endpoint/method/tag pattern. 

- -E (--endpoint) Example: `-E /users/create`, will only test endpoints containing `/users/create`
- -M (--method) Example: `-M GET`, will only test endpoints with GET requests
- -T (--tag) Example: `-T custom_tag`, will only test endpoints with the schema defined tag of `custom_tag`
- -O (--operation-id) Example: `-O custom_operation_id`, will only test the endpoint with the `custom_operation_id` operation_id

Options when generating test data for schema definitions without schema defined examples. When there is no example defined in the API definition, these options will determine how the mock-data is generated when testing that endpoint.

- --hypothesis-derandomize: this flag is used to generate test data deterministically instead of generating random, valid test data.
- --hypothesis-max-examples: this value determines the maximum number of tests to generate for each method. Example: `--hypothesis-max-examples=50`.
- --hypothesis-seed: provide a seed from which random test data will be generated.

### Replay

The `replay` command runs a snapshot of tests and compares new results to the original results. This command takes a file with logs from a previous run of tests.

    ibm-service-validator replay path/to/logs.yaml [options]

Note: To capture these logs, run the service validator with the `--store-request-log` option.

Example: 
- `ibm-service-validator run path/to/mock_server.yaml --base-url https://localhost:5000 --store-request-log=request_log.yaml`
- `ibm-service-validator replay request_log.yaml --status=FAILURE`

This will re-run the service validator on the previous run but only on the endpoints that returned a FAILURE and will report the new result for each endpoint.

#### replay options

- --id: run a specific request by providing the request id (`--id 1`).
- --status: run only the requests with a specific status code (`--status SUCCESS`).
  - SUCCESS
  - ERROR
  - FAILURE
- --uri: run the requests that contain the given uri (`--uri path1/resources`)
- --method: run only the requests that use the given HTTP method (`--method GET`)

## Configuration

### Configuration File

Rules may be configured using a `ibm-service-validator-config.yaml` or `ibm-service-validator-config.json` file.

We look for a configuration file in the current working directory from which the service validator is invoked, and we search up the directory until we find the first matching config file. It is recommended to create this configuration file in the root directory of your project.

Rules may be on, off, or warn. An example of the configuration file is given below:

    ibm_cloud_api_handbook:
        allow_header_in_405: 'on'
        content_location: warn
        get_with_request_body: 'on'
        invalid_accept_header: 'on'
        invalid_request_content_type: 'on'
        location_201: 'on'
        no_422: 'on'
        no_accept_header: 'on'
        no_content_204: 'on'
        www_authenticate_401: 'on'
    schemathesis_checks:
        not_a_server_error: 'on'
        status_code_conformance: 'warn'
        content_type_conformance: 'warn'
        response_schema_conformance: 'on'

### Create Default Configuration File

To initialize a default configuration file in the current working directory, use:

    ibm-service-validator init [options]

#### init options

- -o (--overwrite): if config file already exists, overwrite the existing config file with default values.
- -j (--json): write the default config file as json.

### Add Case Rules

For some rules, we send an additional request to the API to target specific behavior. We call these rules `add_case` rules. The `invalid_request_content_type` rule, for example, tests how the API responds when it receives a request with a request body and an invalid `Content-Type` header. To make this rule effective, we need to send a request to the API with an invalid `Content-Type` header. For `invalid_request_content_type` and all other `add_case` rules listed below, we create an additional request to exercise the behavior the rule is testing. When an `add_case` rule is disabled, the additional request is also disabled.

`add_case` Rules:

- get_with_request_body
- invalid_accept_header
- invalid_request_content_type

## Including Examples in API Definition

Often, a service's requirements are stricter than its schema. For example, an `account_id` may have schema, `type: string`. However, a valid `account_id` is restricted to the set of strings associated with an account. For this reason, the default way to generate requests is to use [OpenAPI examples](https://swagger.io/docs/specification/adding-examples/) in the API definition. Notice examples may be provided using the `example` and `examples` keywords. The service validator supports both `example` and `examples`.

### Important Note on Providing an Object Example

Examples for object properties should be provided in an example object instead of individual property examples. Hence, for the schema:

    schema:
        type: object
        properties:
        id:
            type: integer
        name:
            type: string

Provide an example object:

    schema:
        type: object
        properties:
        id:
            type: integer
        name:
            type: string
        example:
            id: 10
            name: Jessica Smith

Instead of individual property examples:

    schema:
        type: object
        properties:
        id:
            type: integer
            example: 10
        name:
            type: string
            example: Jessica Smith


CHANGELOG.md
# [0.5.0](https://github.com/IBM/service-validator/compare/v0.4.2...v0.5.0) (2020-09-15)


### Features

* Filter tests by operation_id ([a88689a](https://github.com/IBM/service-validator/commit/a88689a34b5609b028be72ced20b59a1e87115ad))

## [0.4.2](https://github.com/IBM/service-validator/compare/v0.4.1...v0.4.2) (2020-08-13)


### Bug Fixes

* remove invalid classifier from setup.py ([#3](https://github.com/IBM/service-validator/issues/3)) ([f8146e2](https://github.com/IBM/service-validator/commit/f8146e263db7d283911b3055b4b267b6f4127909))
* update link to issues and bug reports ([7631422](https://github.com/IBM/service-validator/commit/76314229d6f21288bab63fd27a687136b2fdeac1))

## [0.4.1](https://github.com/IBM/service-validator/compare/v0.4.0...v0.4.1) (2020-08-13)


### Bug Fixes

* update setup.py and pyproject for PyPI distributions ([#2](https://github.com/IBM/service-validator/issues/2)) ([692d31d](https://github.com/IBM/service-validator/commit/692d31d89a761352cca39e43e1cf594be22e4d16))

# [0.4.0](https://github.ibm.com/CloudEngineering/ibm-service-validator/compare/v0.3.0...v0.4.0) (2020-08-11)


### Features

* include links to the API handbook in error messages ([#35](https://github.ibm.com/CloudEngineering/ibm-service-validator/issues/35)) ([c83cb44](https://github.ibm.com/CloudEngineering/ibm-service-validator/commit/c83cb445907326c624ab8b9e0d3fcb8804a6fb23))

# [0.3.0](https://github.ibm.com/CloudEngineering/ibm-service-validator/compare/v0.2.0...v0.3.0) (2020-08-11)


### Features

* conditional checks ([#33](https://github.ibm.com/CloudEngineering/ibm-service-validator/issues/33)) ([f017996](https://github.ibm.com/CloudEngineering/ibm-service-validator/commit/f0179965c3e5c1a97b96c2bc9b418b62de5944b8))

# [0.2.0](https://github.ibm.com/CloudEngineering/ibm-service-validator/compare/v0.1.0...v0.2.0) (2020-06-24)


### Features

* verbosity option ([#28](https://github.ibm.com/CloudEngineering/ibm-service-validator/issues/28)) ([15bdfb9](https://github.ibm.com/CloudEngineering/ibm-service-validator/commit/15bdfb9bdeb1802017b0e521b0ecd16c25d1a2f6))

# [0.1.0](https://github.ibm.com/CloudEngineering/ibm-service-validator/compare/v0.0.2...v0.1.0) (2020-06-23)


### Bug Fixes

* fix distribution file ([fa50283](https://github.ibm.com/CloudEngineering/ibm-service-validator/commit/fa502834b89536d2831ceef065fccf7cf18e57eb))


### Features

* stateful custom test generation ([07f12c0](https://github.ibm.com/CloudEngineering/ibm-service-validator/commit/07f12c02dd66df605014267566defd5342bfeacb))

## [0.0.2](https://github.ibm.com/CloudEngineering/ibm-service-validator/compare/v0.0.1...v0.0.2) (2020-06-11)


### Bug Fixes

* semantic-release ([#24](https://github.ibm.com/CloudEngineering/ibm-service-validator/issues/24)) ([32fbb40](https://github.ibm.com/CloudEngineering/ibm-service-validator/commit/32fbb404c27cb38df24996b3d4c1b2925d7aeb97))


src/ibm_service_validator/cli/__init__.py
#!/usr/bin/env python
# Copyright 2020 IBM All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Callable, Dict, FrozenSet, Iterable, List, Optional, Tuple, Union
import os

import click
import hypothesis
from ibm_cloud_sdk_core.authenticators.iam_authenticator import IAMAuthenticator

from requests.models import Response
from schemathesis.cli.context import ExecutionContext
from schemathesis.cli.handlers import EventHandler
from schemathesis.cli.output.default import DefaultOutputStyleHandler
from schemathesis.cli.output.short import ShortOutputStyleHandler
from schemathesis.cli.options import CSVOption, OptionalInt, NotSet
from schemathesis.cli import callbacks, execute
from schemathesis.cli import replay as _replay
from schemathesis import loaders, runner

from schemathesis.hooks import GLOBAL_HOOK_DISPATCHER, HookContext
from schemathesis.types import Filter
from schemathesis import checks as checks_module
from schemathesis.models import Case

from ibm_service_validator.cli.handlers.output_handler import OutputHandler
from ibm_service_validator.handbook_rules import ADD_CASE_HOOKS, HANDBOOK_RULES
from ibm_service_validator.cli.process_config import (
    create_default_config,
    process_config,
)

ALL_CHECKS: tuple = checks_module.ALL_CHECKS + HANDBOOK_RULES
API_KEY = "IBM_CLOUD_SERVICE_VALIDATOR_API_KEY"
IAM_ENDPOINT = "IBM_CLOUD_SERVICE_VALIDATOR_IAM_ENDPOINT"
DEFAULT_WORKERS: int = 1


@click.group()
@click.option(
    "--set-api-key", "api_key", type=str, help="Set API key as an environment variable."
)
@click.option(
    "--set-iam-endpoint",
    "iam_endpoint",
    type=str,
    help="Set the IAM endpoint as an environment variable.",
)
def ibm_service_validator(api_key: str = str(), iam_endpoint: str = str()) -> None:
    set_environment_variable(api_key, API_KEY)
    set_environment_variable(iam_endpoint, IAM_ENDPOINT)


def set_environment_variable(val: str, env_var_name: str) -> None:
    if val:
        os.environ[env_var_name] = val


@ibm_service_validator.command(short_help="Run a suite of tests.")
@click.argument("schema", type=str, callback=callbacks.validate_schema)
@click.option(
    "--auth",
    "-a",
    type=str,
    callback=callbacks.validate_auth,
    help="Server user and password. Example: USER:PASSWORD",
)
@click.option(
    "--auth-type",
    "-A",
    type=click.Choice(["basic", "digest"], case_sensitive=False),
    default="basic",
    help="The authentication mechanism to be used. Defaults to 'basic'.",
)
@click.option(
    "--base-url",
    "-b",
    callback=callbacks.validate_base_url,
    help="The base-url of the API.",
)
@click.option(
    "--checks",
    "-c",
    type=str,
    default="",
    callback=lambda _, __, s: [c.strip() for c in s.split(",") if c.strip()],
    help="Comma-separated list of checks to run.",
)
@click.option(
    "--header",
    "-H",
    "headers",
    multiple=True,
    type=str,
    callback=callbacks.validate_headers,
    help="Custom header will be used in all requests to server. Ex: Authorization: Bearer 123",
)
@click.option(
    "--endpoint",
    "-E",
    "endpoints",
    type=str,
    multiple=True,
    callback=callbacks.validate_regex,
    help="Filter schemathesis test by endpoint pattern.",
)
@click.option(
    "--exitfirst",
    "-x",
    "exit_first",
    is_flag=True,
    default=False,
    help="Exit instantly on first error or failed test.",
)
@click.option(
    "--hypothesis-deadline",
    # max value to avoid overflow. It is maximum amount of days in milliseconds.
    type=OptionalInt(1, 999999999 * 24 * 3600 * 1000),
    help="Duration in milliseconds each individual example is not allowed to exceed.",
)
@click.option(
    "--hypothesis-derandomize",
    help="Use Hypothesis's deterministic mode.",
    is_flag=True,
    default=None,
)
@click.option(
    "--hypothesis-max-examples",
    type=click.IntRange(1),
    help="Maximum number of generated examples per each method/endpoint combination.",
)
@click.option(
    "--hypothesis-phases",
    default="explicit",
    type=CSVOption(hypothesis.Phase),
    help="Control which phases should be run.",
)
@click.option(
    "--hypothesis-seed", type=int, help="Set a seed to use for all Hypothesis tests."
)
@click.option(
    "--hypothesis-verbosity",
    type=click.Choice([item.name for item in hypothesis.Verbosity]),
    callback=callbacks.convert_verbosity,
    help="Verbosity level of Hypothesis messages.",
)
@click.option(
    "--method",
    "-M",
    "methods",
    type=str,
    multiple=True,
    callback=callbacks.validate_regex,
    help="Filter schemathesis test by HTTP method.",
)
@click.option(
    "--no-additional-cases",
    is_flag=True,
    default=False,
    help="Additional requests that target specific API behavior will not be sent.",
)
@click.option(
    "--request-timeout",
    type=click.IntRange(1),
    help="Timeout in milliseconds for network requests during the test run.",
)
@click.option(
    "--show-exception-tracebacks",
    is_flag=True,
    default=False,
    help="Show full tracebacks for internal exceptions.",
)
@click.option(
    "--statistics",
    "-s",
    is_flag=True,
    default=False,
    help="Show statistical summary of errors.",
)
@click.option(
    "--store-request-log",
    help="Store requests and responses into a file.",
    type=click.File("w"),
)
@click.option(
    "--tag",
    "-T",
    "tags",
    type=str,
    multiple=True,
    callback=callbacks.validate_regex,
    help="Filter schemathesis test by schema tag pattern.",
)
@click.option(
    "--operation-id",
    "-O",
    "operation_ids",
    type=str,
    multiple=True,
    help="Filter schemathesis test by operationId pattern.",
    callback=callbacks.validate_regex,
)
@click.option(
    "--validate-schema",
    help="Enable or disable validation of input schema.",
    type=bool,
    default=True,
)
@click.option("--verbosity", "-v", help="Control verbosity of output.", count=True)
@click.option(
    "--with-bearer", "-B", is_flag=True, help="Flag to send bearer token with requests."
)
def run(  # pylint: disable=too-many-arguments
    schema: str,
    auth: Optional[Tuple[str, str]],
    auth_type: str,
    base_url: Optional[str],
    checks: Optional[List[str]],
    headers: Dict[str, str],
    hypothesis_phases: Optional[List[hypothesis.Phase]],
    endpoints: Optional[Filter] = None,
    exit_first: bool = False,
    hypothesis_deadline: Optional[Union[int, NotSet]] = None,
    hypothesis_derandomize: Optional[bool] = None,
    hypothesis_max_examples: Optional[int] = None,
    hypothesis_seed: Optional[int] = None,
    hypothesis_verbosity: Optional[hypothesis.Verbosity] = None,
    methods: Optional[Filter] = None,
    no_additional_cases: bool = False,
    request_timeout: Optional[int] = None,
    show_exception_tracebacks: bool = False,
    statistics: bool = False,
    store_request_log: Optional[click.utils.LazyFile] = None,
    tags: Optional[Filter] = None,
    operation_ids: Optional[Filter] = None,
    validate_schema: bool = True,
    verbosity: int = 0,
    with_bearer: bool = False,
) -> None:
    # pylint: disable=too-many-locals

    if with_bearer:
        if "Authorization" not in headers and "authorization" not in headers:
            headers["Authorization"] = get_bearer_token()
        else:
            raise click.UsageError(
                "--with-bearer flag used but Authorization header provided with --header."
            )

    on, warnings = (frozenset(checks), frozenset()) if checks else process_config()

    selected_checks = get_selected_checks(on)
    register_output_handler(warnings, statistics)
    if not no_additional_cases:
        register_add_case_hooks(on)

    # Invoke Schemathesis
    prepared_runner = runner.prepare(
        schema,
        app=None,
        auth=auth,
        auth_type=auth_type,
        base_url=base_url,
        checks=selected_checks,
        endpoint=endpoints,
        exit_first=exit_first,
        headers=headers,
        loader=loaders.from_path,
        method=methods,
        request_timeout=request_timeout,
        seed=hypothesis_seed,
        store_interactions=store_request_log is not None,
        tag=tags,
        operation_id=operation_ids,
        validate_schema=validate_schema,
        workers_num=DEFAULT_WORKERS,
        hypothesis_deadline=hypothesis_deadline,
        hypothesis_derandomize=hypothesis_derandomize,
        hypothesis_max_examples=hypothesis_max_examples,
        hypothesis_phases=hypothesis_phases,
        hypothesis_report_multiple_bugs=False,
        hypothesis_suppress_health_check=None,
        hypothesis_verbosity=hypothesis_verbosity,
    )
    execute(
        prepared_runner,
        DEFAULT_WORKERS,
        show_exception_tracebacks,
        store_request_log,
        None,
        verbosity,
    )


def get_selected_checks(
    on: FrozenSet[str],
) -> Iterable[Callable[[Response, Case], None]]:
    return tuple(check for check in ALL_CHECKS if check.__name__ in on)


def register_output_handler(warnings: FrozenSet[str], statistics: bool) -> None:
    def after_init_cli_run_handlers(
        context: HookContext,
        handlers: List[EventHandler],
        execution_context: ExecutionContext,
    ) -> None:
        # filters out the Schemathesis output handlers and adds OutputHandler
        handlers[:] = [
            *filter(
                lambda handler: not isinstance(handler, DefaultOutputStyleHandler)
                and not isinstance(handler, ShortOutputStyleHandler),
                handlers,
            ),
            OutputHandler(warnings, statistics),
        ]

    GLOBAL_HOOK_DISPATCHER.register(after_init_cli_run_handlers)


def register_add_case_hooks(on: FrozenSet) -> None:
    # pylint: disable=bad-str-strip-call
    add_case_prefix = "add_"
    for case_hook in ADD_CASE_HOOKS:
        # add add_case hook if its corresponding check is on
        if case_hook.__name__.lstrip(add_case_prefix) in on:
            GLOBAL_HOOK_DISPATCHER.register_hook_with_name(case_hook, "add_case")


def get_bearer_token() -> str:
    if API_KEY not in os.environ or IAM_ENDPOINT not in os.environ:
        raise click.UsageError(
            f"Must set {API_KEY} and {IAM_ENDPOINT} environment variables to use --with-bearer."
        )

    try:
        bearer_token = IAMAuthenticator(
            os.environ[API_KEY], url=os.environ[IAM_ENDPOINT]
        ).token_manager.get_token()
        return "Bearer {0}".format(bearer_token)
    except Exception as e:  # pragma: no cover
        raise RuntimeError("Problem getting bearer token.") from e  # pragma: no cover


@ibm_service_validator.command(short_help="Create a default config file.")
@click.option(
    "--overwrite",
    "-o",
    "overwrite",
    is_flag=True,
    default=False,
    help="Overwrite an existing config file with the default config.",
)
@click.option(
    "--json",
    "-j",
    "write_json",
    is_flag=True,
    default=False,
    help="Overwrite an existing config file with the default config.",
)
def init(write_json: bool, overwrite: bool) -> None:
    create_default_config(write_json, overwrite)


@ibm_service_validator.command(short_help="Replay requests from a saved request log.")
@click.argument("cassette_path", type=click.Path(exists=True))
@click.option("--id", "id_", help="ID of request to replay.", type=str)
@click.option(
    "--status", help="Status (ERROR, FAILURE, SUCCESS) of requests to replay.", type=str
)
@click.option(
    "--uri", help="A regexp that filters requests by their request URI.", type=str
)
@click.option(
    "--method",
    help="A regexp that filters requests by their request method.",
    type=str,
)
@click.pass_context
def replay(  # pylint: disable=too-many-arguments
    context: click.Context,
    cassette_path: str,
    id_: Optional[str],
    status: Optional[str],
    uri: Optional[str],
    method: Optional[str],
) -> None:
    # pylint: disable=too-many-locals
    context.forward(_replay)
